File: platform-common/src/main/java/org/sweetchips/platform/common/Workflow.java
package org.sweetchips.platform.common;

import org.sweetchips.utility.ItemsUtil;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Future;
import java.util.concurrent.FutureTask;
import java.util.concurrent.RunnableFuture;
import java.util.stream.Collectors;

public final class Workflow {

    private static final String TAG = "Workflow";

    private final ContextLogger mLogger;
    private List<Collection<RootUnit>> mWorkSet = new ArrayList<>();
    private List<Runnable> mPrepareBefore = new ArrayList<>();
    private List<Runnable> mPrepareAfter = new ArrayList<>();
    private List<Runnable> mTransformBefore = new ArrayList<>();
    private List<Runnable> mTransformAfter = new ArrayList<>();

    public Workflow(ContextLogger logger) {
        mLogger = logger;
    }

    public void apply(PlatformContext context) {
        addPrepareBefore(context.onPrepareBefore());
        addPrepareAfter(context.onPrepareAfter());
        addTransformBefore(context.onTransformBefore());
        addTransformAfter(context.onTransformAfter());
    }

    public void addWork(Collection<RootUnit> collection) {
        ItemsUtil.checkAndAdd(mWorkSet, collection);
    }

    public Future<?> start(Executor executor) {
        mLogger.d(TAG, "start: begin");
        List<Collection<RootUnit>> list = mWorkSet;
        if (list == null) {
            throw new IllegalStateException();
        }
        mWorkSet = null;
        List<Transformer> transformers = list.stream().map(Transformer::new).collect(Collectors.toList());
        Runnable runnable = () -> Transformer.Companion.doWork$platform_common(transformers,
                this::prepareBefore,
                this::prepareAfter,
                this::transformBefore,
                this::transformAfter,
                mLogger);
        RunnableFuture<?> future = new FutureTask<>(runnable, null);
        executor.execute(future);
        mLogger.d(TAG, "start: end");
        return future;
    }

    public void addPrepareBefore(Runnable runnable) {
        ItemsUtil.checkAndAdd(mPrepareBefore, runnable);
    }

    public void addPrepareAfter(Runnable runnable) {
        ItemsUtil.checkAndAdd(mPrepareAfter, runnable);
    }

    public void addTransformBefore(Runnable runnable) {
        ItemsUtil.checkAndAdd(mTransformBefore, runnable);
    }

    public void addTransformAfter(Runnable runnable) {
        ItemsUtil.checkAndAdd(mTransformAfter, runnable);
    }

    private void prepareBefore() {
        mLogger.d(TAG, "prepareBefore: begin");
        mPrepareBefore.forEach(Runnable::run);
        mPrepareBefore = null;
        mLogger.d(TAG, "prepareBefore: end");
    }

    private void prepareAfter() {
        mLogger.d(TAG, "prepareAfter: begin");
        mPrepareAfter.forEach(Runnable::run);
        mPrepareAfter = null;
        mLogger.d(TAG, "prepareAfter: end");
    }

    private void transformBefore() {
        mLogger.d(TAG, "transformBefore: begin");
        mTransformBefore.forEach(Runnable::run);
        mTransformBefore = null;
        mLogger.d(TAG, "transformBefore: end");
    }

    private void transformAfter() {
        mLogger.d(TAG, "transformAfter: begin");
        mTransformAfter.forEach(Runnable::run);
        mTransformAfter = null;
        mLogger.d(TAG, "transformAfter: end");
    }
}

File: platform-common/src/main/kotlin/org/sweetchips/platform/common/Workflow.kt
package org.sweetchips.platform.common

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.supervisorScope
import org.sweetchips.utility.FilesUtil
import java.util.concurrent.Executor
import java.util.concurrent.Future
import java.util.concurrent.FutureTask

class Workflow(private val logger: ContextLogger) {

    private companion object {

        private const val TAG = "Workflow"

        suspend fun Collection<RootUnit>.doPrepare() = supervisorScope {
            forEach {
                when (it.status) {
                    RootUnit.Status.REMOVED, RootUnit.Status.NOTCHANGED -> Unit
                    RootUnit.Status.ADDED, RootUnit.Status.CHANGED -> launch {
                        it.onPrepare()
                    }
                }
            }
        }

        suspend fun Collection<RootUnit>.doTransform() = supervisorScope {
            forEach {
                when (it.status) {
                    RootUnit.Status.REMOVED -> launch(Dispatchers.IO) {
                        FilesUtil.deleteIfExists(it.output)
                    }
                    RootUnit.Status.NOTCHANGED -> Unit
                    RootUnit.Status.ADDED, RootUnit.Status.CHANGED -> launch {
                        it.onTransform()
                    }
                }
            }
        }
    }

    private var workSet: MutableList<Collection<RootUnit>>? = mutableListOf()

    private var prepareBefore: MutableList<Runnable>? = mutableListOf()
    private var prepareAfter: MutableList<Runnable>? = mutableListOf()
    private var transformBefore: MutableList<Runnable>? = mutableListOf()
    private var transformAfter: MutableList<Runnable>? = mutableListOf()

    fun addWork(collection: Collection<RootUnit>) = workSet?.add(collection)

    fun addPrepareBefore(runnable: Runnable) = prepareBefore?.add(runnable)
    fun addPrepareAfter(runnable: Runnable) = prepareAfter?.add(runnable)
    fun addTransformBefore(runnable: Runnable) = transformBefore?.add(runnable)
    fun addTransformAfter(runnable: Runnable) = transformAfter?.add(runnable)

    fun attach(context: PlatformContext) {
        addPrepareBefore(context.onPrepareBefore())
        addPrepareAfter(context.onPrepareAfter())
        addTransformBefore(context.onTransformBefore())
        addTransformAfter(context.onTransformAfter())
    }

    fun start(executor: Executor): Future<Unit> {
        logger.d(TAG, "start: begin")
        val future = workSet!!.also { workSet = null }.let {
            FutureTask { it.doRun() }
        }
        executor.execute(future)
        logger.d(TAG, "start: end")
        return future
    }

    private fun Collection<Collection<RootUnit>>.doRun(
    ) {
        logger.d(TAG, "doWork: begin")
        prepareBefore!!.also { prepareBefore = null }.phase("prepareBefore")
        doWork("doPrepare") { doPrepare() }
        prepareAfter!!.also { prepareBefore = null }.phase("prepareAfter")
        transformBefore!!.also { prepareBefore = null }.phase("transformBefore")
        doWork("doTransform") { doTransform() }
        transformAfter!!.also { prepareBefore = null }.phase("transformAfter")
        logger.d(TAG, "doWork: end")
    }

    private fun List<Runnable>.phase(tag: String) {
        logger.d(TAG, "$tag: begin")
        forEach { it.run() }
        logger.d(TAG, "$tag: end")
    }

    private fun Collection<Collection<RootUnit>>.doWork(
        tag: String,
        work: suspend Collection<RootUnit>.() -> Unit
    ) {
        logger.d(TAG, "$tag: begin")
        runBlocking {
            forEach {
                launch { it.work() }
            }
        }
        logger.d(TAG, "$tag: end")
    }
}